### PR TITLE
fixed #16938: Audio could not be played entirely on iOS/macOS. And refactored some code of decoding.

### LINKED ISSUE #16938: Audio could not be played entirely on iOS/macOS
Reported at http://forum.cocos.com/t/topic/41723

- cocos2d-x version: v3.14
- devices test on: iOS, macOS
- developing environments
   - NDK version: N/A
   - Xcode version: 8.1
   - VS version: N/A
   - browser type and version:N/A

Steps to Reproduce:

Just play this audio [1018.mp3.zip](https://github.com/cocos2d/cocos2d-x/files/621300/1018.mp3.zip).

There are some frames at the end could not be played.
It seems that audio automatically stops before it's finished.

### PR DESCRIPTION
**Pull request by dumganhar**

### CODE REVIEW

File: cocos/audio/apple/AudioDecoder.h [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+     * @brief Opens an audio file specified by a file path.
+     * @return true if succeed, otherwise false.
+     */
+    bool open(const char* path);
+
+    /**
+     * @brief Checks whether decoder has opened file successfully.
+     * @return true if succeed, otherwise false.
+     */
+    bool isOpened();
```
Comment (minggo) [inline cocos/audio/apple/AudioDecoder.h]: add const?

Comment (minggo) [reply cocos/audio/apple/AudioDecoder.h]: and please check other functions

Comment (dumganhar) [reply cocos/audio/apple/AudioDecoder.h]: I forgot this.  I'll do it.

Comment (dumganhar) [reply cocos/audio/apple/AudioDecoder.h]: DONE. I will add a test case for this issue later.

**Pull request merged**